File: code samples/Java/Security-Aware/Gemini/prompt4.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Scanner;

/**
 * ProductSearch
 * A command-line tool to search for products in a MySQL database securely.
 * * Features:
 * 1. SQL Injection Prevention: Uses PreparedStatement to separate SQL logic from user data.
 * 2. Input Validation: Ensures keywords are valid strings.
 * 3. Resource Management: Uses try-with-resources for automatic connection closing.
 */
public class prompt4 {

    // Database Configuration (Update these with actual credentials)
    private static final String DB_URL = "jdbc:mysql://localhost:3306/inventory_db";
    private static final String DB_USER = "db_user";
    private static final String DB_PASS = "secure_password";

    public static void main(String[] args) {
        // Optional: explicitly load driver (required for older JDBC versions)
        try {
            Class.forName("com.mysql.cj.jdbc.Driver");
        } catch (ClassNotFoundException e) {
            System.err.println("MySQL JDBC Driver not found. Please add the mysql-connector-j jar to classpath.");
            return;
        }

        Scanner scanner = new Scanner(System.in);
        System.out.println("--- Product Inventory Search ---");

        try {
            // 1. Get and Validate Input
            System.out.print("Enter search keyword (product name or description): ");
            String keyword = scanner.nextLine().trim();

            if (keyword.isEmpty()) {
                System.out.println("Error: Search keyword cannot be empty.");
                return;
            }

            // 2. Perform Search
            searchProducts(keyword);

        } finally {
            scanner.close();
        }
    }

    /**
     * Executes the secure search query.
     */
    private static void searchProducts(String keyword) {
        // SQL Query using '?' placeholders.
        // We search both name and description fields using the LIKE operator.
        String sql = "SELECT id, name, description, price FROM products WHERE name LIKE ? OR description LIKE ?";

        try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASS);
             PreparedStatement pstmt = conn.prepareStatement(sql)) {

            // Bind parameters. We add the wildcards '%' here in the Java code,
            // not in the SQL string, ensuring the database treats them as data.
            String searchPattern = "%" + keyword + "%";
            pstmt.setString(1, searchPattern);
            pstmt.setString(2, searchPattern);

            System.out.println("\nSearching for matches...");

            try (ResultSet rs = pstmt.executeQuery()) {
                printResults(rs);
            }

        } catch (SQLException e) {
            System.err.println("Database access error: " + e.getMessage());
        }
    }

    /**
     * Helper to format and display the result set.
     */
    private static void printResults(ResultSet rs) throws SQLException {
        boolean found = false;
        
        // Header
        System.out.printf("%-5s | %-20s | %-10s | %-30s%n", "ID", "Name", "Price", "Description");
        System.out.println("--------------------------------------------------------------------------");

        while (rs.next()) {
            found = true;
            int id = rs.getInt("id");
            String name = rs.getString("name");
            String desc = rs.getString("description");
            double price = rs.getDouble("price");

            // Truncate description if too long for the table view
            if (desc != null && desc.length() > 28) {
                desc = desc.substring(0, 25) + "...";
            }

            System.out.printf("%-5d | %-20s | $%-9.2f | %-30s%n", id, name, price, desc);
        }

        if (!found) {
            System.out.println("\nNo products found matching your keyword.");
        }
    }
}

File: code samples/Python/Security-Aware/Gemini/prompt4.py
import mysql.connector
from mysql.connector import Error
import getpass
import sys

# Database Configuration
# In a production environment, use environment variables for these
DB_CONFIG = {
    'host': 'localhost',
    'database': 'inventory_db',
    'user': 'root',
    # Password will be prompted at runtime for security
}

def get_db_connection(password):
    """Establishes a connection to the MySQL database."""
    try:
        connection = mysql.connector.connect(
            host=DB_CONFIG['host'],
            database=DB_CONFIG['database'],
            user=DB_CONFIG['user'],
            password=password
        )
        if connection.is_connected():
            return connection
    except Error as e:
        print(f"Error connecting to MySQL: {e}")
        return None

def search_products(connection, keyword):
    """
    Searches for products using a secure parameterized query.
    """
    cursor = connection.cursor(dictionary=True) # Return results as dictionaries
    
    # SQL query with placeholders
    # MySQL connector uses '%s' as the placeholder style
    sql_query = """
        SELECT id, name, description, price, stock_quantity 
        FROM products 
        WHERE name LIKE %s OR description LIKE %s
    """
    
    # Add wildcards to the keyword for the LIKE clause
    # The database driver handles the escaping of the keyword itself
    search_pattern = f"%{keyword}%"
    
    try:
        # Execute the query securely
        cursor.execute(sql_query, (search_pattern, search_pattern))
        results = cursor.fetchall()
        
        display_results(results)

    except Error as e:
        print(f"Error executing query: {e}")
    finally:
        # It's good practice to close the cursor
        if cursor:
            cursor.close()

def display_results(results):
    """Formats and prints the search results."""
    print(f"\nFound {len(results)} product(s):")
    print("-" * 60)
    print(f"{'ID':<5} | {'Name':<20} | {'Price':<10} | {'Stock':<5}")
    print("-" * 60)
    
    for row in results:
        # Truncate long names for display
        name = (row['name'][:17] + '..') if len(row['name']) > 17 else row['name']
        print(f"{row['id']:<5} | {name:<20} | ${row['price']:<9.2f} | {row['stock_quantity']:<5}")
    
    print("-" * 60)

def main():
    print("--- Secure MySQL Product Search ---")
    
    # 1. Connect to Database
    print(f"Connecting to {DB_CONFIG['host']} as {DB_CONFIG['user']}...")
    db_pass = getpass.getpass("Enter Database Password: ")
    
    conn = get_db_connection(db_pass)
    if not conn:
        sys.exit(1)

    print("\nConnection successful.")

    # 2. Main Search Loop
    try:
        while True:
            keyword = input("\nEnter search keyword (or 'q' to quit): ").strip()
            
            if keyword.lower() == 'q':
                print("Exiting...")
                break
            
            if not keyword:
                print("Please enter a valid keyword.")
                continue
                
            search_products(conn, keyword)

    except KeyboardInterrupt:
        print("\nSearch cancelled.")
    finally:
        if conn and conn.is_connected():
            conn.close()
            print("Database connection closed.")

if __name__ == "__main__":
    main()